processor/src/main/java/org/mapstruct/ap/internal/model/source/TargetReference.java

```diff
@@ -142,46 +173,150 @@ public TargetReference build() {
+                    // there should always be a write accessor (except for the last when the mapping is ignored and
+                    // there is a read accessor) and there should be read accessor mandatory for all but the last
+                    if ( reportError ) {
+                        //TODO maybe it would be better if we store the errors in the builder and do a report in the
+                        // end, so not iterate this twice
+                        reportError( targetWriteAccessor, targetReadAccessor );
+                    }
                     break;
                 }
 
-                if ( ( i == entryNames.length - 1 ) || ( Executables.isSetterMethod( targetWriteAccessor )
+                if ( isLast || ( Executables.isSetterMethod( targetWriteAccessor )
                     || Executables.isFieldAccessor( targetWriteAccessor ) ) ) {
                     // only intermediate nested properties when they are a true setter or field accessor
                     // the last may be other readAccessor (setter / getter / adder).
 
-                    if ( Executables.isGetterMethod( targetWriteAccessor ) ||
-                        Executables.isFieldAccessor( targetWriteAccessor ) ) {
-                        nextType = typeFactory.getReturnType(
-                            (DeclaredType) nextType.getTypeMirror(),
-                            targetWriteAccessor );
-                    }
-                    else {
-                        nextType = typeFactory.getSingleParameter(
-                            (DeclaredType) nextType.getTypeMirror(),
-                            targetWriteAccessor ).getType();
-                    }
+                    nextType = findNextType( nextType, targetWriteAccessor, targetReadAccessor, isLast );
 
                     // check if an entry alread exists, otherwise create
                     String[] fullName = Arrays.copyOfRange( entryNames, 0, i + 1 );
                     PropertyEntry propertyEntry = PropertyEntry.forTargetReference( fullName, targetReadAccessor,
                         targetWriteAccessor, nextType );
                     targetEntries.add( propertyEntry );
-                    }
-
                 }
 
+            }
+
             return targetEntries;
         }
 
+        /**
+         * Finds the next type based on the initial type.
+         *
+         * @param initial for which a next type should be found
+         * @param targetWriteAccessor the write accessor
+         * @param targetReadAccessor the read accessor
+         * @param isLast whether this is the call for the last accessor
+         *
+         * @return the next type that should be used for finding a property entry
+         */
+        private Type findNextType(Type initial, Accessor targetWriteAccessor,
+            Accessor targetReadAccessor, boolean isLast) {
+            if ( !isLast && targetWriteAccessor == null ) {
```

Comment:
We don't really need to check it. If this condition is `true` then we would have already stopped on line 188. The condition is unreachable unless we change something in the lines 178 to 180. I mostly added during implementation in case I mess something up. If you think that we should remove it, I can do it.